src/context/WalletProvider/WalletProvider.tsx

```diff
@@ -730,104 +787,92 @@ export const WalletProvider = ({ children }: { children: React.ReactNode }): JSX
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [state.adapters, state.keyring])
 
-  const handleAccountsOrChainChanged = useCallback(async () => {
-    if (!walletType || !state.adapters) return
+  const handleAccountsOrChainChanged = useCallback(
+    async (localWalletType: KeyManagerWithProvider | null, accountsOrChains: string[] | string) => {
+      if (!localWalletType || !state.adapters) return
 
-    const adapter = await getAdapter(walletType)
-    const localWallet = await adapter?.pairDevice()
+      // Note, we NEED to use store.getState instead of the walletType variable above
+      // The reason is handleAccountsOrChainChanged exists in the context of a closure, hence will keep a stale reference forever
+      const _walletType = store.getState().localWalletSlice.walletType
 
-    if (!localWallet) return
+      // This shouldn't happen if event listeners are properly removed, but they may not be
+      // This fixes the case of switching from e.g MM, to another wallet, then switching accounts/chains in MM and MM becoming connected again
+      if (_walletType && localWalletType !== _walletType) return
 
-    await localWallet.initialize()
-    const deviceId = await localWallet?.getDeviceID()
+      const _isLocked = Array.isArray(accountsOrChains) && accountsOrChains.length === 0
```

Comment:
- `chainChanged` fires an event with a chain as a `string`
- `accountsChanged` fires an event with the new accounts as `string[]`
- MM going to locked state fires `accountsChanged` with an empty accounts list, whereas unlocking it will fire it with the currently connected accounts' list